### PR TITLE
[VolvoOnCall] OH3 update

### PR DESCRIPTION
**Pull request by clinique**
Solving issue #8554  and issue #8518
Some code corrections and enhancements.
Introduced new trigger channels (CAR_STOPPED and CAR_MOVED)
Initiated French translation.

### CODE REVIEW

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/api/ActionResultControler.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link ActionResultControler} is responsible for triggering information
+ * update after a post has been submitted to the webservice.
+ *
+ * @author Gaël L'hopital - Initial contribution
+ */
+@NonNullByDefault
+public class ActionResultControler implements Runnable {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/api/ActionResultControler.java:34]:
```suggestion
public class ActionResultController implements Runnable {
```

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/api/ActionResultControler.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    private final Logger logger = LoggerFactory.getLogger(ActionResultControler.class);
+
+    private final VocHttpApi service;
+    private final ScheduledExecutorService scheduler;
+    private PostResponse postResponse;
```
Comment (cpmeister) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/api/ActionResultControler.java:39]:
```suggestion
    private final PostResponse postResponse;
```

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/api/ActionResultControler.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    private ThingHandler vehicle;
```
Comment (cpmeister) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/api/ActionResultControler.java:41]:
```suggestion
    private final ThingHandler vehicle;
```

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/api/ActionResultControler.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        this.vehicle = vehicle;
+    }
+
+    @Override
+    public void run() {
+        switch (postResponse.status) {
+            case SUCCESSFULL:
+            case FAILED:
+                logger.info("Action {} for vehicle {} resulted : {}.", postResponse.serviceType.toString(),
+                        postResponse.vehicleId, postResponse.status.toString());
```
Comment (cpmeister) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/api/ActionResultControler.java:57]:
```suggestion
                logger.debug("Action {} for vehicle {} resulted : {}.", postResponse.serviceType,
                        postResponse.vehicleId, postResponse.status);
```

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/api/ActionResultControler.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                vehicle.handleCommand(vehicle.getThing().getChannels().get(0).getUID(), RefreshType.REFRESH);
+                break;
+            default:
+                try {
+                    postResponse = service.getURL(postResponse.serviceURL, PostResponse.class);
+                    scheduler.schedule(this, 10000, TimeUnit.MILLISECONDS);
+                } catch (VolvoOnCallException e) {
+                    if (e.getType() == ErrorType.SERVICE_UNAVAILABLE || e.getType() == ErrorType.INTERRUPTED) {
+                        scheduler.schedule(this, 10000, TimeUnit.MILLISECONDS);
```
Comment (cpmeister) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/api/ActionResultControler.java:66]: If you received an interrupt you shouldn't schedule more things. Typically an interrupt indicates a shutdown is occurring.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/dto/HvBattery.java [hunk 6]
```diff
@@ -26,7 +27,7 @@
 public class HvBattery {
     public int hvBatteryLevel = UNDEFINED;
     public int distanceToHVBatteryEmpty = UNDEFINED;
-    public @NonNullByDefault({}) String hvBatteryChargeStatusDerived;
+    public @NonNullByDefault({}) StringType hvBatteryChargeStatusDerived;
```
Comment (cpmeister) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/dto/HvBattery.java:30]: If this actually works then I have no problem with this change.

File: bundles/org.openhab.binding.volvooncall/src/main/resources/OH-INF/thing/vehicle.xml [hunk 7]
```diff
@@ -264,8 +266,8 @@
 	<channel-type id="fuelConsumption" advanced="true">
 		<item-type>Number</item-type>
 		<label>Average Consumption</label>
-		<description>Indicates the average fuel consumption in L/100km</description>
-		<state pattern="%.1f L/100km" readOnly="true"></state>
+		<description>Indicates the average fuel consumption in l/100km</description>
+		<state pattern="%.1f l/100km" readOnly="true"></state>
```
Comment (cpmeister) [inline bundles/org.openhab.binding.volvooncall/src/main/resources/OH-INF/thing/vehicle.xml:270]: Why are the units in volume/distance instance of distance/volume? Is this just a Volvo thing?

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VolvoOnCallBridgeHandler.java [hunk 8]
```diff
@@ -58,144 +42,60 @@
...
+            CustomerAccounts account = api.getURL("customeraccounts/", CustomerAccounts.class);
+            if (account.username != null) {
+                updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE, account.username);
             } else {
-                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
-                        "Incorrect username or password");
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Incorrect login credentials");
             }
-        } catch (JsonSyntaxException | VolvoOnCallException e) {
+        } catch (Exception e) {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VolvoOnCallBridgeHandler.java:71]: Please catch specific exceptions, or at the very least catch RuntimeException instead.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java [hunk 9]
```diff
@@ -80,20 +82,77 @@
...
+                        }
+
+                        updateStatus(ThingStatus.ONLINE);
+                        startAutomaticRefresh(configuration.refresh, api);
+                    } catch (VolvoOnCallException e) {
+                        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.HANDLER_INITIALIZING_ERROR, e.getMessage());
+                    }
+
+                }
+                ;
```
Comment (cpmeister) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/handler/VehicleHandler.java:145]: orphan semicolon

**Review by cpmeister — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/dto/HvBattery.java (hunk 6 above)
Comment (clinique) [reply bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/dto/HvBattery.java:30]: Yes, it works, it's taken in charge by gson jsonDeserializer declaration.

File: bundles/org.openhab.binding.volvooncall/src/main/resources/OH-INF/thing/vehicle.xml (hunk 7 above)
Comment (clinique) [reply bundles/org.openhab.binding.volvooncall/src/main/resources/OH-INF/thing/vehicle.xml:270]: Not especially Volvo, it more European to indicate fuel consumption / 100km - I saw that in US (or the rest of the world) it's different.

File: bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/api/VocHttpApi.java [hunk 10]
```diff
@@ -60,14 +60,19 @@
...
         this.cache = new ExpiringCacheMap<>(120 * 1000);
         this.configuration = configuration;
         this.httpClient = httpClient;
 
         httpClient.setUserAgentField(new HttpField(HttpHeader.USER_AGENT, "openhab/voc_binding/" + InstanceUUID.get()));
-        httpClient.start();
+        try {
+            httpClient.start();
+        } catch (Exception e) {
+            throw new VolvoOnCallException(new IOException("Unable to start Jetty HttpClient"));
```
Comment (cpmeister) [inline bundles/org.openhab.binding.volvooncall/src/main/java/org/openhab/binding/volvooncall/internal/api/VocHttpApi.java:74]: Don't lose track of the source exception.
```suggestion
            throw new VolvoOnCallException(new IOException("Unable to start Jetty HttpClient", e));
```

**Review by cpmeister — CHANGES_REQUESTED** [review]

**Review by cpmeister — APPROVED** [review]

**Pull request merged**